src/main/kotlin/org/jetbrains/intellij/platform/gradle/tasks/TestIdeTask.kt

```diff
@@ -98,8 +98,48 @@ abstract class TestIdeTask : Test(), TestableAware, IntelliJPlatformVersionAware
                     .toSet()
             })
 
-            classpath = project.fileTree(sourceTask.sandboxPluginsDirectory) + instrumentedTestCode + classpath + productModules - mainSourceSet
+            // The below is needed to simulate the behavior of com.intellij.ide.plugins.cl.PluginClassLoader
+            // which is present in the IDE when the plugin is used in "production".
+            // https://plugins.jetbrains.com/docs/intellij/plugin-class-loaders.html
+            //
+            // The main difference is that the plugin classloader will first try to load any class from the libraries
+            // present in the plugin and only then delegate the loading to the IDE.
+            //
+            // This way the libs, which are present in the "lib" directory of the plugin's zip distribution file can
+            // override IDE's dependencies.
+            //
+            // But when we run tests, the classpath is built by Gradle according to the defined dependencies, and there
+            // is no separation into "plugin's dependencies" and the "IDE's dependencies"; it is one big list.
+            //
+            // So to make the test environment more like the production environment, we should put the plugin's direct
+            // dependencies the first on the classpath. That way the class loading works effectively the same as in
+            // the production with the plugin classloader.
+
+            // Removes the default runtime dependencies from the classpath because their order is unknown.
+            classpath -= sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME).runtimeClasspath
+
+            // Adds back the same dependencies, but it takes them instead from the plugin's test sandbox.
+            // Taking them from the sandbox is more correct because they might be slightly different there,
+            // e.g. the plugin might be doing some customizations while creating the distribution zip file,
+            // like obfuscating some code or patching libs etc.
+            classpath = sourceTask.sandboxPluginsDirectory.asFileTree.filter {
+                // Filters the files to keep only jar files to avoid breaking the JVM.
+                // It is necessary because some plugins have arbitrary files in their distribution zip file,
+                // which break the JVM when added to the classpath.
+                // TODO: Should we take libs only from "lib" dir or all libs present in the plugin ".zip" file?
```

Comment:
The Ant plugin also has "dist/ant.jar" which will get on the classpath without additional filtering.